### PR TITLE
Adds a filter Aggregator.

### PR DESCRIPTION
**Pull request by MadVikingGod**
There is an ongoing discussion in #3011 about the best way to solve when and where to filter attributes.  This does not resolve that issue.

This is just a solution to enable us to get to alpha.  This puts the computation on the ingest side (solution 1 in #3011), this path was taken so that each aggregator doesn't need to have some logic combining the DataPoints during aggregation.

The intended use is when creating the aggregator for a pipeline, the SDK will wrap the aggregator in a filter.

### CODE REVIEW

File: sdk/metric/internal/filter.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+// Aggregate records the measurement, scoped by attr, and aggregates it
+// into an aggregation.
+func (f *filter[N]) Aggregate(measurement N, attr attribute.Set) {
+	fAttr, ok := f.seen[attr]
+	if !ok {
+		fAttr = f.filter(attr)
+		f.seen[attr] = fAttr
+	}
+	f.aggregator.Aggregate(measurement, fAttr)
```
Comment (MrAlias) [inline sdk/metric/internal/filter.go:53]: This method will need to be concurrent safe. This should be guarded by a lock.

**Review by MrAlias — COMMENTED** [review]
This is missing tests to validate the functionality introduced here.

File: sdk/metric/internal/filter.go (hunk 1 above)
Comment (MadVikingGod) [reply sdk/metric/internal/filter.go:53]: You're right, I will add it.

File: sdk/metric/internal/filter.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+// Aggregation returns an Aggregation, for all the aggregated
+// measurements made and ends an aggregation cycle.
+func (f *filter[N]) Aggregation() metricdata.Aggregation {
+	return f.aggregator.Aggregation()
```
Comment (MrAlias) [inline sdk/metric/internal/filter.go:64]: The intrinsic unbounded nature of `seen` and its propensity to grow indefinitely is troubling. Do we want to add clearing code here? Possibly some iteration over the returned `Aggregation` and clearing the map of unused attributes?

If we do not want to include this in this PR, can we add a TODO and reference https://github.com/open-telemetry/opentelemetry-go/issues/3006 (or its own issue?).

**Review by MrAlias — APPROVED** [review]

**Review by jmacd — APPROVED** [review]

**Review by dashpole — APPROVED** [review]

**Pull request merged**